libcst/_nodes/internal.py

```diff
@@ -70,6 +71,30 @@ def record_syntactic_position(
         yield
 
 
+@add_slots
+@dataclass(frozen=False)
+class CodegenWriter(CodegenState):
+    """
+    A CodegenState that writes to a file-like object.
+    """
+
+    writer: io.TextIOBase = None  # need a default value for dataclass
+
+    def __post_init__(self) -> None:
+        if self.writer is None:
+            raise TypeError("writer must be provided")
+
+    def add_indent_tokens(self) -> None:
+        for token in self.indent_tokens:
+            self.writer.write(token)
+
+    def add_token(self, value: str) -> None:
+        self.writer.write(value)
+
+    def pop_trailing_newline(self) -> None:
+        pass
```

Comment:
this is incorrect, code that doesn't end in a trailing newline won't be serialized properly (for example the python module `""` will be serialized as `"\n"`)